html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php

```diff
@@ -0,0 +1,151 @@
+    return new static(
+      $container->get('entity.repository'),
+      $container->get('file_system'),
+      $container->get('path_alias.manager'),
+    );
+  }
+
+  /**
+   * Generate the download file and serves it to the browser.
+   *
+   * @param \Drupal\node\NodeInterface $node
+   *   The node.
+   * @param string $format
+   *   The format of the file to serve.
+   *
+   * @return Symfony\Component\HttpFoundation\BinaryFileResponse
+   *   The file to download.
+   */
+  public function createFile(NodeInterface $node, string $format): BinaryFileResponse {
+    // Not Found if $format is not a supported file extension.
+    if (!in_array($format, static::SUPPORTED_EXTENSIONS, TRUE)) {
+      throw new NotFoundHttpException();
+    }
+
+    $paragraph_field_items = $node->get('field_columns')->referencedEntities();
+    foreach ($paragraph_field_items as $paragraph) {
+      // Get the translation.
+      $paragraph = $this->entityRepository->getTranslationFromContext($paragraph);
+    }
+    $sheetData = [
+      [
+        'column_allowed_values',
+        'column_data_quality',
+        'column_description',
+        'column_name',
+        'column_size',
+        'column_transformations',
+        'provenance_field_name',
+        'provenance_field_number',
+        'provenance_field_question',
+        'provenance_form_name',
+        'provenance_form_number',
+      ],
+      [
+        $paragraph->get('field_column_allowed_values')->value,
+        $paragraph->get('field_column_data_quality')->value,
+        $paragraph->get('field_column_description')->value,
+        $paragraph->get('field_column_name')->value,
+        $paragraph->get('field_column_size')->value,
+        $paragraph->get('field_column_transformations')->value,
+        $paragraph->get('field_provenance_field_name')->value,
+        $paragraph->get('field_provenance_field_number')->value,
+        $paragraph->get('field_provenance_field_question')->value,
+        $paragraph->get('field_provenance_form_name')->value,
+        $paragraph->get('field_provenance_form_number')->value,
+      ],
+    ];
+
+    // Make a filename like "node-file-path_ID_12.csv".
+    $node_path = $this->pathAliasManager->getAliasByPath('/node/' . $node->id());
+    $node_path = basename($node_path);
+    $filename = $node_path . '_ID_' . $node->id() . '.' . $format;
+
+    // Create a file path to the temp directory. PhpSpreadsheet does not work
+    // with stream wrappers.
+    $file_path = $this->fileSystem->getTempDirectory() . '/' . $filename;
+
+    // Generate the spreadsheet object and save to $file_path.
+    $spreadsheet = new Spreadsheet();
+    $worksheets = new Worksheet($spreadsheet);
+    $spreadsheet->addSheet($worksheets);
+    $worksheets->fromArray($sheetData);
+    $sheetIndex = $spreadsheet->getIndex(
+      $spreadsheet->getSheetByName('Worksheet')
+    );
+    $spreadsheet->removeSheetByIndex($sheetIndex);
```

Comment:
Why is this needed? Does a new spreadsheet come with a blank worksheet that needs to be removed? If so, a comment to this effect would be good.

Reply:
Comment had been added